lib/semantics/unparse-with-symbols.cc

```diff
@@ -0,0 +1,154 @@
+  symbolMap refs_;  // statement location to symbol referenced there
+  std::map<const Symbol *, const char *> symbolToStmt_;  // symbol to def
+
+  void EndScope();
+  bool PreReference();
+  void PostReference();
+  bool isRef() const { return isRef_ > 0; }
+  void PrintSymbols(const parser::CharBlock &, std::ostream &, int,
+      const symbolMap &, bool) const;
+  void Indent(std::ostream &, int) const;
+};
+
+void SymbolDumpVisitor::PrintSymbols(
+    const parser::CharBlock &location, std::ostream &out, int indent) const {
+  PrintSymbols(location, out, indent, defs_, true);
+  PrintSymbols(location, out, indent, refs_, false);
+}
+void SymbolDumpVisitor::PrintSymbols(const parser::CharBlock &location,
+    std::ostream &out, int indent, const symbolMap &symbols, bool isDef) const {
+  std::set<const Symbol *> done;  // used to prevent duplicates
+  auto range{symbols.equal_range(location.begin())};
+  for (auto it{range.first}; it != range.second; ++it) {
+    auto *symbol{it->second};
+    if (done.insert(symbol).second) {
+      Indent(out, indent);
+      out << '!' << (isDef ? "DEF"s : "REF"s) << ": ";
+      DumpForUnparse(out, symbol->GetUltimate(), isDef);
+      out << '\n';
+    }
+  }
+}
+void SymbolDumpVisitor::Indent(std::ostream &out, int indent) const {
+  for (int i{0}; i < indent; ++i) {
+    out << ' ';
+  }
+}
+
+void SymbolDumpVisitor::Post(const parser::Name &name) {
+  if (const auto *symbol{name.symbol}) {
+    CHECK(currStmt_);
+    // If this is the first reference to an implicitly defined symbol,
+    // record it as a def.
+    bool isImplicit{symbol->test(Symbol::Flag::Implicit) &&
+        symbolToStmt_.find(symbol) == symbolToStmt_.end()};
+    if (isRef() && !isImplicit) {
+      refs_.emplace(currStmt_->begin(), symbol);
+    } else {
+      symbolToStmt_.emplace(symbol, currStmt_->begin());
+    }
+  }
+}
+
+// Defs are initially saved in symbolToStmt_ so that a symbol defined across
+// multiple statements is associated with only one (the first). Now that we
+// are at the end of a scope, move them into defs_.
+void SymbolDumpVisitor::EndScope() {
+  for (auto pair : symbolToStmt_) {
+    defs_.emplace(pair.second, pair.first);
+  }
+  symbolToStmt_.clear();
+}
+
+// {Pre,Post}Reference() are called around constructs that contains symbols
+// references. Sometimes those are nested (e.g. DataRef inside Designator)
+// so we need to maintain a count to know when we are back out.
+bool SymbolDumpVisitor::PreReference() {
+  ++isRef_;
+  return true;
+}
+void SymbolDumpVisitor::PostReference() {
+  CHECK(isRef_ > 0);
+  --isRef_;
+}
+
+void UnparseWithSymbols(std::ostream &out, const parser::Program &program,
+    parser::Encoding encoding) {
+  SymbolDumpVisitor visitor;
+  parser::Walk(program, visitor);
+  parser::preStatementType preStatement{
+      [&](const parser::CharBlock &location, std::ostream &out, int indent) {
+        visitor.PrintSymbols(location, out, indent);
+      }};
```

Comment:
It should be possible to define `preStatement` as a pointer to the member function with `std:bind`, but my attempt to do so didn't compile.